File: Klein-Gordon/useless functions.cpp
#include <iostream>
#include <vector>
#include <cmath>
#include <random>
using namespace std;

auto r = mt19937 {123};
auto stand = uniform_real_distribution<double>{0., 1.};

void mse_err(const vector<double>& vec, const double& mean, double& mse, double& err)
{
  int i;
  int N = vec.size();

  double var = 0.;
  for (i = 0; i < N; i ++) var += pow(vec[i] - mean, 2.);
  var /= double(N-1);

  double result = 0.;
  for (i = 0; i < N; i ++) result += pow(vec[i] - mean, 4.);

  mse = var / double(N);
  err = pow(((result/double(N)) - pow(var, 2.)*double(N-3)/double(N-1)) / pow(double(N), 3.), .5);
}

class Field2d
{
  private:
    int L1_, L2_;
    double am_, κ_, κ2_;
    vector<double> data_;
    int index_(int x, int y) const { return x + L1_*y; }
  public:
    Field2d(int L1, int L2, double am) : L1_(L1), L2_(L2), am_(am), κ_(pow(4. + am_*am_, -.5)), κ2_(1. / (4. + am_*am_)), data_(L1_*L2_)
    {
      int L_total = L1_*L2_;
      for (int i = 0; i < L_total; i ++) data_[i] = 0.;
    }

    double& operator() (int x, int y)       { return data_[index_(x, y)]; }
    double  operator() (int x, int y) const { return data_[index_(x, y)]; }
    int L1() const { return L1_; }
    int L2() const { return L2_; }
    double am() const { return am_; }
    double kappa() const { return κ_; }
    double kappa2() const { return κ2_; }
};

void Gibbs(Field2d& φ)
{
  int i, j;
  double γ;
  for (j = 0; j < φ.L2(); j ++)
  {
    for (i = 0; i < φ.L1(); i ++)
    {
      γ = φ((i+1) % φ.L1(), j) + φ(i, (j+1) % φ.L2()) + φ((i-1+φ.L1()) % φ.L1(), j) + φ(i, (j-1+φ.L2()) % φ.L2());
      auto norm = normal_distribution<double>{φ.kappa2()*γ, φ.kappa()};
      φ(i, j) = norm(r);
    }
  }
}

void Metropolis(Field2d& φ, const auto& dist)
{
  int i, j;
  double curr, prop, κ2γ, diff;
  for (j = 0; j < φ.L2(); j ++)
  {
    for (i = 0; i < φ.L1(); i ++)
    {
      curr = φ(i, j);
      prop = curr + dist(r);
      κ2γ = φ.kappa2() * (φ((i+1) % φ.L1(), j) + φ(i, (j+1) % φ.L2()) + φ((i-1+φ.L1()) % φ.L1(), j) + φ(i, (j-1+φ.L2()) % φ.L2()));
      diff = pow(curr - κ2γ, 2.) - pow(prop - κ2γ, 2.);
      if (diff >= 0. || stand(r) < exp(diff / (2.*φ.kappa2()))) φ(i, j) = prop;
    }
  }
}

void GibbsCorrelation(Field2d& φ, const int& N, const int& disc, vector<vector<double>>& c_means, vector<vector<vector<double>>>& c_mses, vector<vector<vector<double>>>& c_mse_errs)
{
  for (int j = 0; j < φ.L2(); j ++)
  {
    for (int i = 0; i < φ.L1(); i ++) φ(i, j) = 0.;
  }

  for (int n = 0; n < disc; n ++)
  {
    Gibbs(φ);
    if (n % 100000 == 0) cout << n << "\n";
  }

  int K = int(.5 + log2(double(N)));

  vector<vector<double>> Φs(2);
  Φs[0] = vector<double>(φ.L1());
  Φs[1] = vector<double>(φ.L2());

  // FIX L -> L+1 FOR CS VECTOR AND LOOPS
  vector<vector<vector<double>>> cs(2);
  cs[0] = vector<vector<double>>(φ.L1());
  cs[1] = vector<vector<double>>(φ.L2());
  for (int i = 0; i < φ.L1(); i ++) cs[0][i] = vector<double>(N);
  for (int i = 0; i < φ.L2(); i ++) cs[1][i] = vector<double>(N);

  for (int i = 0; i < φ.L1(); i ++)
  {
    Φs[0][i] = 0.;
    for (int j = 0; j < N; j ++) cs[0][i][j] = 0.;
    for (int j = 0; j < K; j ++)
    {
      c_mses[0][i][j] = 0.;
      c_mse_errs[0][i][j] = 0.;
    }
  }
  for (int i = 0; i < φ.L1()+1; i ++) c_means[0][i] = 0.;
  for (int i = 0; i < φ.L2(); i ++)
  {
    Φs[1][i] = 0.;
    for (int j = 0; j < N; j ++) cs[1][i][j] = 0.;
    for (int j = 0; j < K; j ++)
    {
      c_mses[1][i][j] = 0.;
      c_mse_errs[1][i][j] = 0.;
    }
  }
  for (int i = 0; i < φ.L2()+1; i ++) c_means[1][i] = 0.;

  for (int n = 0; n < N; n ++)
  {
    Gibbs(φ);
    // SET PHIS TO BE 0 EACH ITERATION BEFORE SUMMING
    for (int i = 0; i < φ.L1(); i ++)
    {
      for (int j = 0; i < φ.L2(); j ++)
      {
        Φs[0][i] += φ(i, j);
        Φs[1][j] += φ(i, j);
      }
    }

    for (int i = 0; i < φ.L1(); i ++)
    {
      for (int j = 0; j <= φ.L1(); j ++)
      {
        cs[0][j][n] += Φs[0][i] * Φs[0][(i+j) % φ.L1()];
        c_means[0][j] += cs[0][j][n];
      }
    }
    for (int i = 0; i < φ.L2(); i ++)
    {
      for (int j = 0; j <= φ.L2(); j ++)
      {
        cs[1][j][n] += Φs[1][i] * Φs[1][(i+j) % φ.L2()];
        c_means[1][j] += cs[1][j][n];
      }
    }

    if (n % 100000 == 0) cout << n << "\n";
  }

  for (int i = 0; i < φ.L1()+1; i ++)
  {
    c_means[0][i] /= double(N);

    int bins = N;
    vector<double> c(bins);
    for (int n = 0; n < bins; n ++) c[n] = cs[0][i][n];
    mse_err(c, c_means[0][i], c_mses[0][i][0], c_mse_errs[0][i][0]);

    for (int j = 1; j < K; j ++)
    {
      bins /= 2;
      for (int n = 0; n < bins; n ++) c[n] = .5 * (c[2*n] + c[2*n + 1]);
      c.resize(bins);
      mse_err(c, c_means[0][i], c_mses[0][i][j], c_mse_errs[0][i][j]);
    }
  }
  for (int i = 0; i < φ.L2()+1; i ++)
  {
    c_means[1][i] /= double(N);

    int bins = N;
    vector<double> c(bins);
    for (int n = 0; n < bins; n ++) c[n] = cs[1][i][n];
    mse_err(c, c_means[1][i], c_mses[1][i][0], c_mse_errs[1][i][0]);

    for (int j = 1; j < K; j ++)
    {
      bins /= 2;
      for (int n = 0; n < bins; n ++) c[n] = .5 * (c[2*n] + c[2*n + 1]);
      c.resize(bins);
      mse_err(c, c_means[1][i], c_mses[1][i][j], c_mse_errs[1][i][j]);
    }
  }
}

File: Klein-Gordon/misc/useless functions.py
import numpy as np
import statistics as stat


def mse_err(array):
    """Mean squared error and mean squared error error.

    For a given array, returns the mean squared error and its corresponding
    error for the sample mean estimator.
    """
    mean, var, M = stat.fmean(array), stat.variance(array), len(array)
    return (var/M,
            (((stat.fmean((array-mean)**4.))
              - ((var**2.) * (M-3.))/(M-1.)) / (M**3.)) ** .5)


def twodsquareGibbsCorrelation(L, am, N, disc):
    κ = (4. + (am) ** 2.) ** (-0.5)
    lattice = np.zeros((L, L))
    for n in range(disc):
        for i in range(L):
            for j in range(L):
                γ = (lattice[(i+1) % L, j]
                     + lattice[i, (j+1) % L]
                     + lattice[(i-1) % L, j]
                     + lattice[i, (j-1) % L])
                η = np.random.normal(0., 1.)
                lattice[i, j] = κ * (κ*γ + η)
        if n % 100000 == 0:
            print(n)
    Φs = np.zeros((2, L))
    K = int(np.log2(N))
    cs = np.zeros((N, 2, L//2 + 1))
    c_mses, c_mse_errs = (np.empty((2, L//2 + 1, K)) for _ in range(2))
    for n in range(N):
        for i in range(L):
            for j in range(L):
                γ = (lattice[(i+1) % L, j]
                     + lattice[i, (j+1) % L]
                     + lattice[(i-1) % L, j]
                     + lattice[i, (j-1) % L])
                η = np.random.normal(0., 1.)
                lattice[i, j] = κ * (κ*γ + η)
        for i in range(L):
            Φs[0, i] = np.sum(lattice[i])
            Φs[1, i] = np.sum(lattice[:, i])
        for d in range(2):
            for δ in range(L//2 + 1):
                for x in range(L):
                    cs[n, d, δ] += Φs[d, x] * Φs[d, (x+δ) % L]
        if n % 100000 == 0:
            print(n)
    c_means = np.mean(cs, axis=0)
    for d in range(2):
        for δ in range(L//2 + 1):
            c = cs[:, d, δ]
            c_mses[d, δ, 0], c_mse_errs[d, δ, 0] = mse_err(c)
            for k in range(1, K):
                bins = len(c) // 2
                binned = np.empty(bins)
                for b in range(bins):
                    binned[b] = .5 * (c[2*b] + c[2*b + 1])
                c = binned
                c_mses[d, δ, k], c_mse_errs[d, δ, k] = mse_err(c)
            print(d, δ)
    return c_means, c_mses, c_mse_errs
    # return c_means, c_mses, c_mse_errs, cs[0, 0, ::10]


def twodsquareMetropolisCorrelation(L, am, ε, N, disc):
    κ2 = 1. / (4. + am ** 2.)
    lattice = np.zeros((L, L))
    for n in range(disc):
        for i in range(L):
            for j in range(L):
                curr = lattice[i, j]
                κ2γ = κ2 * (lattice[(i+1) % L, j]
                            + lattice[i, (j+1) % L]
                            + lattice[(i-1) % L, j]
                            + lattice[i, (j-1) % L])
                prop = curr + np.random.uniform(-ε, ε)
                diff = (curr - κ2γ) ** 2. - (prop - κ2γ) ** 2.
                if (diff >= 0.) or (
                        np.random.uniform(0, 1) < np.exp(diff / (2.*κ2))):
                    lattice[i, j] = prop
        if n % 100000 == 0:
            print(n)
    Φs = np.zeros((2, L))
    K = int(np.log2(N))
    cs = np.zeros((N, 2, L//2 + 1))
    c_mses, c_mse_errs = (np.empty((2, L//2 + 1, K)) for _ in range(2))
    for n in range(N):
        for i in range(L):
            for j in range(L):
                curr = lattice[i, j]
                κ2γ = κ2 * (lattice[(i+1) % L, j]
                            + lattice[i, (j+1) % L]
                            + lattice[(i-1) % L, j]
                            + lattice[i, (j-1) % L])
                prop = curr + np.random.uniform(-ε, ε)
                diff = (curr - κ2γ) ** 2. - (prop - κ2γ) ** 2.
                if (diff >= 0.) or (
                        np.random.uniform(0, 1) < np.exp(diff / (2.*κ2))):
                    lattice[i, j] = prop
        for i in range(L):
            Φs[0, i] = np.sum(lattice[i])
            Φs[1, i] = np.sum(lattice[:, i])
        for d in range(2):
            for x in range(L):
                for δ in range(L+1):
                    cs[d, δ, n] += Φs[d, x] * Φs[d, (x+δ) % L]
        if n % 100000 == 0:
            print(n)
    c_means = np.mean(cs, axis=0)
    for d in range(2):
        for δ in range(L//2 + 1):
            c = cs[:, d, δ]
            c_mses[d, δ, 0], c_mse_errs[d, δ, 0] = mse_err(c)
            for k in range(1, K):
                bins = len(c) // 2
                binned = np.empty(bins)
                for b in range(bins):
                    binned[b] = .5 * (c[2*b] + c[2*b + 1])
                c = binned
                c_mses[d, δ, k], c_mse_errs[d, δ, k] = mse_err(c)
            print(d, δ)
    return c_means, c_mses, c_mse_errs


def twodsquareGibbs(length, am, N):
    """Gibbs sampler.

    Performs Gibbs sampling for a Klein-Gordon field over a 2D square
    lattice for a given lattice size, mass parameter, and number of iterations,
    and returns the sum of the field over each dimension at each iteration.
    """
    κ = (4. + (am) ** 2.) ** (-0.5)
    lattice = np.zeros((length, length))
    Φs = np.zeros((N, 2, length))
    for n in range(N):
        for i in range(length):
            for j in range(length):
                γ = (lattice[(i+1) % length, j]
                     + lattice[i, (j+1) % length]
                     + lattice[(i-1) % length, j]
                     + lattice[i, (j-1) % length])
                η = np.random.normal(0., 1.)
                lattice[i, j] = κ * (κ*γ + η)
        for i in range(length):
            Φs[n, 0, i] = np.sum(lattice[i])
        for j in range(length):
            Φs[n, 1, j] = np.sum(lattice[:, j])
        if n % 10000. == 0.:
            print(n)
    return Φs


def twodsquareCorrelation(Φs):
    """Sitewise correlation and the binning method.

    Estimates the sitewise correlation and performs the binning method for a
    given array of field sums, returning the estimated sitewise correlation
    and the binned mean squared errors (and its corresponding error). Requires
    the number of iterations to be a power of 2 for the binning method.
    """
    N, length = np.shape(Φs)[0], np.shape(Φs)[2]
    K = int(np.log2(N))
    c_means = np.empty((2, length+1))
    c_mses, c_mse_errs = (np.empty((2, K, length+1)) for _ in range(2))
    for d in range(2):
        cs = np.zeros((N, length+1))
        for n in range(N):
            for x in range(length):
                for δ in range(length+1):
                    cs[n, δ] += Φs[n, d, x] * Φs[n, d, (x+δ) % length]
            if n % 10000 == 0:
                print(d, n)
        c_means[d] = np.mean(cs, axis=0)
        for δ in range(length+1):
            c = cs[:, δ]
            c_mses[d, 0, δ], c_mse_errs[d, 0, δ] = mse_err(c)
            for k in range(1, K):
                bins = int(len(c) / 2)
                binned = np.empty(bins)
                for b in range(bins):
                    binned[b] = .5 * (c[2*b] + c[2*b + 1])
                c = binned
                c_mses[d, k, δ], c_mse_errs[d, k, δ] = mse_err(c)
            print(d, δ)
    return c_means, c_mses, c_mse_errs


def Gibbs(shape, am, N):
    """Gibbs simulation of a scalar field over a lattice (periodic)."""
    D = len(shape)
    L_total = 1
    for k in shape:
        L_total *= k
    κ = (2.*D + (am)**2.) ** (-0.5)
    lattice = np.zeros(shape)
    Φs = np.empty((N, D), dtype=object)
    for n in range(N):
        indices = np.zeros(D, dtype=int)
        for x in range(L_total):
            γ = 0.
            for d in range(D):
                L = shape[d]
                indices_ = np.delete(indices, d)
                index = indices[d]
                lattice_ = np.moveaxis(lattice, d, -1)[tuple(indices_)]
                γ += lattice_[(index-1) % L] + lattice_[(index+1) % L]
            η = np.random.normal(0., 1.)
            lattice[tuple(indices)] = κ * (κ*γ + η)

            indices[0] += 1
            for d in range(D-1):
                if indices[d] == shape[d]:
                    indices[d] = 0
                    indices[d+1] += 1

        for d in range(D):
            L = shape[d]
            Φs[n, d] = np.empty(L)
            lattice_ = np.moveaxis(lattice, d, 0)
            for x in range(L):
                Φs[n, d][x] = np.sum(lattice_[x])
        if n % 100. == 0.:
            print(n)

    return Φs


def Metropolis(shape, am, N, Δ):
    """Metropolis simulation of a scalar field over a lattice (periodic)."""
    D = len(shape)
    L_total = 1
    for k in shape:
        L_total *= k
    κ2 = 1. / (2.*D + (am)**2.)
    lattice = np.zeros(shape)
    Φs = np.empty((N, D), dtype=object)
    for n in range(N):
        indices = np.zeros(D, dtype=int)
        for x in range(L_total):
            γ = 0.
            for d in range(D):
                L = shape[d]
                indices_ = np.delete(indices, d)
                index = indices[d]
                lattice_ = np.moveaxis(lattice, d, -1)[tuple(indices_)]
                γ += lattice_[(index-1) % L] + lattice_[(index+1) % L]
            φ = lattice[tuple(indices)]
            φ_ = φ + np.random.uniform(-Δ, Δ)
            ψ, ψ_ = (φ - κ2*γ) ** 2., (φ_ - κ2*γ) ** 2.
            if ψ_ <= ψ or np.random.uniform(0, 1) < np.exp((ψ - ψ_) / (2.*κ2)):
                lattice[tuple(indices)] = φ_

            indices[0] += 1
            for d in range(D-1):
                if indices[d] == shape[d]:
                    indices[d] = 0
                    indices[d+1] += 1

        for d in range(D):
            L = shape[d]
            Φs[n, d] = np.empty(L)
            lattice_ = np.moveaxis(lattice, d, 0)
            for x in range(L):
                Φs[n, d][x] = np.sum(lattice_[x])
        if n % 100. == 0.:
            print(n)

    return Φs


def Correlation(Φs):
    N, D = np.shape(Φs)
    K = int(np.log2(N))
    c_means = np.empty(D, dtype=object)
    c_mses, c_mse_errs = (np.empty((D, K), dtype=object) for _ in range(2))
    for d in range(D):
        L = len(Φs[0, d])
        cs = np.zeros((N, L+1))
        for n in range(N):
            for x in range(L):
                for δ in range(L+1):
                    cs[n, δ] += Φs[n, d][x] * Φs[n, d][(x+δ) % L]
        c_means[d] = np.mean(cs, axis=0)
        for k in range(K):
            c_mses[d, k], c_mse_errs[d, k] = (np.empty(L+1) for _ in range(2))
        for δ in range(L+1):
            c = cs[:, δ]
            c_mses[d, 0][δ], c_mse_errs[d, 0][δ] = mse_err(c)
            for k in range(1, K):
                bins = int(len(c) / 2)
                binned = np.empty(bins)
                for b in range(bins):
                    binned[b] = .5 * (c[2*b] + c[2*b + 1])
                c = binned
                c_mses[d, k][δ], c_mse_errs[d, k][δ] = mse_err(c)
    return c_means, c_mses, c_mse_errs
